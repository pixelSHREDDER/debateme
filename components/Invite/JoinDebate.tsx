'use client'

import joinDebate from '@/actions/join-debate'
import { useFormState } from 'react-dom'
import { useUser } from '@auth0/nextjs-auth0/client'
import FormSubmitButton from '../Forms/FormSubmitButton'

const initialState = {
  message: '',
}

export default function JoinDebate({ debateId }: { debateId: number }) {
  const { user, error, isLoading } = useUser()
  const [state, formAction] = useFormState(joinDebate, initialState)

  if (isLoading) { return 'loading user...' }
  if (error) { return JSON.stringify(error) }
  if (!user || !user.sub) { return 'please login' }
  if (!debateId) { return 'Debate not found' }

  return (
      <section data-testid="join-section">
          <h2>Join this Debate</h2>
          <p>You have been invited to join this debate. Click below to join.</p>
          <form action={formAction}>
          <input
            type="hidden"
            id="debateId"
            name="debateId"
            value={debateId} />
            <input
              type="hidden"
              id="opponentSub"
              name="opponentSub"
              value={user.sub} />
            <p aria-live="polite">
              {state?.message}
            </p>
            <FormSubmitButton label="Join" />
          </form>
      </section>
    )
}
