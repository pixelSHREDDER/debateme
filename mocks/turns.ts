import { TTurn } from '@/lib/prisma-types'

// Lorem Ipsum HTML courtesy of https://css-tricks.com/new-site-html-ipsum/

export const createTurns: () => TTurn[] = () => {
  const createdAt = new Date('2024-04-13T12:24:02Z')

  return [
    {
      id: 1,
      body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain? <span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span> Ligulas step drops both? You shall not pass!</p>
          <p>Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil. <span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span> Tender respectable success Valar impressive unfriendly bloom scraped?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>
          <hr>`,
      debateId: 1,
      createdAt,
      userSub: 'auth0|1234567890abcdefghij1234',
      },
      {
        id: 2,
        body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p><span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span></p>
          <p>Ligulas step drops both? You shall not pass! Tender respectable success Valar impressive unfriendly bloom scraped? Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil.</p>
          <p><span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span></p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>`,
        debateId: 1,
        createdAt,
        userSub: 'auth0|abcdefghij1234567890abcd',
        },
    ] as const
  }
