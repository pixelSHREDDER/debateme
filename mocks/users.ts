import { UserProfile } from '@auth0/nextjs-auth0/client'

export const getUsers: () => UserProfile[] = () => [
    {
      nickname: 'ron.swanson12345',
      name: 'Ron Swanson',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Frs.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'ron.swanson12345@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|1234567890abcdefghij1234',
      sid: 'o_abcdefgh12345678-ABCDEF12345',
    },
    {
      nickname: 'leslie.knope54321',
      name: 'Leslie Knope',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Flk.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'leslie.knope54321@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|abcdefghij1234567890abcd',
      sid: 'o_12345678abcdefgh-12345ABCDEF',
    },
  ] as const
