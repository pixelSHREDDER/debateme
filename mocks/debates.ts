import { TDebate } from '@/lib/prisma-types'
import { DebateStatus } from '@prisma/client'
import { createTurns } from './turns'

export const createDebates: () => TDebate[] = () => {
  const createdAtDate = new Date('2024-04-13T12:24:02Z')
  const cooldownDate = new Date()
  const createdTurns = createTurns()

  return [
    {
      id: 1,
      topic: 'Your turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      turn: createdTurns,
      status: DebateStatus.CreatorTurn,
    },
    {
      id: 2,
      topic: 'Opponent\'s Turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.OpponentTurn,
      turn: [
        createdTurns[0],
      ],
    },
    {
      id: 3,
      topic: 'Wait for Cooldown',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.CreatorCooldown,
      turn: [
        createdTurns[0],
        {
          ...createdTurns[1],
          createdAt: cooldownDate,
        },
      ],
    },
    {
      id: 4,
      topic: 'No opponent',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
    {
      id: 5,
      topic: 'You\'re Invited to join',
      createdAt: createdAtDate,
      creatorSub: 'auth0|abcdefghij1234567890abcd',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
  ] as const
}
