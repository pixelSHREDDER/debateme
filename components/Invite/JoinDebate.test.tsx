import * as auth0Hooks from '@auth0/nextjs-auth0/client'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import ProviderWrapper from '@/tests/helpers/ProviderWrapper'
import { getUsers } from '@/mocks/users'
import JoinDebate from './JoinDebate'

const users: auth0Hooks.UserProfile[] = getUsers()

describe('Join Debate component', () => {
  const useUserSpy = vi.spyOn(auth0Hooks, 'useUser')

  beforeEach(() => {
    vi.useFakeTimers()
    useUserSpy.mockReturnValue({
      checkSession: vi.fn(),
      isLoading: false,
      user: users[0],
    })
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  describe('Join Debate', () => {
    test('renders correctly', () => {
      render(
        <JoinDebate debateId={4} inviteId="" />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.getByTestId('join-section')).toMatchSnapshot()
    })

    test('does not render if user is loading', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: true,
        user: users[0],
      })
      render(
        <JoinDebate debateId={4} inviteId="" />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('join-section')).toBeNull
    })

    test('does not render if user is not defined', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: false,
      })
      render(
        <JoinDebate debateId={4} inviteId="" />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('join-section')).toBeNull
    })
  })

  describe('Joining debate', () => {
    test('joins debate when button is clicked', async () => {
      render(
        <JoinDebate debateId={4} inviteId="" />,
        { wrapper: ProviderWrapper },
      )
      fireEvent.submit(screen.getByTestId('join-debate-form'))
    })
  })
})
