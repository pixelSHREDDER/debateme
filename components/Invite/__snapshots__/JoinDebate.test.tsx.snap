// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Join Debate component > Join Debate > renders correctly 1`] = `
<section
  data-testid="join-section"
>
  <h2>
    Join this Debate
  </h2>
  <p>
    You have been invited to join this debate. Click below to join.
  </p>
  <form
    data-testid="join-debate-form"
  >
    <input
      id="debateId"
      name="debateId"
      type="hidden"
      value="4"
    />
    <input
      id="inviteId"
      name="inviteId"
      type="hidden"
      value=""
    />
    <input
      id="opponentSub"
      name="opponentSub"
      type="hidden"
      value="auth0|1234567890abcdefghij1234"
    />
    <p
      aria-live="polite"
    />
    <button
      class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
      data-variant="filled"
      style="--button-radius: var(--mantine-radius-md); --button-bg: var(--mantine-color-marble-filled); --button-hover: var(--mantine-color-marble-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; flex-shrink: 0;"
      type="submit"
    >
      <span
        class="m_80f1301b mantine-Button-inner"
      >
        <span
          class="m_811560b9 mantine-Button-label"
        >
          Join
        </span>
      </span>
    </button>
  </form>
</section>
`;
